### ISSUE TITLE
Implicit operator from Result<T> to Result<U>

### ISSUE THREAD

**Issue by PatrizioG** [ardalis/Result#206]
Hi
I would be interested in having an implicit operator from ```Result<T>``` to ```Result<U>``` just to maintain the state of internal properties.
I have a method that can return a ```Result<int>``` on success:
``` csharp
 protected virtual async Task<Result<int>> SaveChangesReturnResultAsync()
 {
     try
     {
         var numberOfWritings = await Context.SaveChangesAsync();
         return Result.Success(numberOfWritings);
     }
     catch (Exception ex)
     {
         if (ex.IsEntityAlreadyExistsException())
             return Result.Conflict("Entity Already Exists");

         if (ex.IsForeignKeyViolation())
             return Result.Conflict("Foreign Key Violation");

         throw;
     }
 }
```

But the caller must be able to return a ```Result<Tentity>``` if successful or propogate errors from the previous call:
``` csharp
  public virtual async Task<Result<TEntity>> CreateAndReturnResultAsync(TEntity entity)
  {
      await Context.AddAsync(entity);
      var result = await SaveChangesReturnResultAsync();
      if (!result.IsSuccess)
          return result; // Conversion error
      return Result.Success(entity);
  }
```

Currently what I have been able to obtain from the code base is a new implicit operator:
``` csharp
public static implicit operator Result(Result<T> result) => new()
{
    Status = result.Status,
    Errors = result.Errors,
    SuccessMessage = result.SuccessMessage,
    CorrelationId = result.CorrelationId,
    ValidationErrors = result.ValidationErrors,
};
```

But this fix would force you to perform two conversion steps:
``` csharp
    public virtual async Task<Result<TEntity>> CreateAndReturnResultAsync(TEntity entity)
    {
        await Context.AddAsync(entity);
        var result = await SaveChangesReturnResultAsync();
        Result temp = result; // From Result<int> to Result, to me it's ok to lost the original value.
        if (!result.IsSuccess)
            return temp; // Conversion ok
        return Result.Success(entity);
    }
```
Thanks in advance

**Comment by jmrtnz94** [comment]
Would the `Map` extensions method fit your use case?

https://github.com/ardalis/Result/blob/main/src/Ardalis.Result/ResultExtensions.cs#L17

**Comment by PatrizioG** [comment]
Thank you very much, now I understand how to use the map method to handle my specific case. I close the issue.

**Comment by PatrizioG** [comment]
But with this approach it seems to me that I can't map from Result<T> to Result keeping internal errors and status.
Any suggestion?

**Comment by KyleMcMaster** [comment]
> But with this approach it seems to me that I can't map from `Result<T>` to `Result` keeping internal errors and status. Any suggestion?

This is fixed in #213, we're working on getting that PR merged soon and a release out shortly.

**Comment by KyleMcMaster** [comment]
@PatrizioG I believe this issue can be marked as closed as release 10.1.0 includes an overload of `Map` that allows converting from `Result<T>` to `Result`. 

Example usage here: 
https://github.com/ardalis/Result/blob/d38ccbd9f9f3f761b8cb8431f461f1837f9ae6ae/tests/Ardalis.Result.UnitTests/ResultGenericToVoidMap.cs#L71-L84

**Comment by PatrizioG** [comment]
Ok @KyleMcMaster thanks so much!
